ADR: Vulnerability scanning modes (#208)
open-component-model/open-delivery-gear: docs/adr/005-vulnerability-scanning-modes.md

# ADR 005: Vulnerability Scanning Modes

| Status   | Proposed                     |
|----------|------------------------------|
| Date     | 2026-08-03                   |
| Deciders | Jonas Brand, Chris Schneider |

## Context and Problem Statement

ADR 004 decided on a scanner-agnostic data model for `VulnerabilityFinding`. This ADR decides

> **what inputs vulnerability scanners receives**: the OCM artefact binary (image, blob, archive) and/or a SBOM already generated for it.

Reviewer: ```suggestion
**which inputs vulnerability scanners receive**: the OCM artefact binary (image, blob, archive) and/or an SBOM already generated for it.
```

ODG already runs a Syft-based SBOM-Generator extension that scans each OCM resource and stores a
CycloneDX SBOM in blob storage. Vulnerability scanners like Trivy and Grype can accept either
an image reference/binary **or** a CycloneDX SBOM as input.

The question is whether a new scanner extension should:

1. Re-fetch and scan the artefact binary directly, or
2. Consume the existing SBOM from OCM component or produced by the SBOM-Generator, or
3. do both

### How OCM Resources Are Accessed

> ODG resources have three access types that scanners must handle:

Reviewer: ```suggestion
ODG currently supports three resource access types that scanners must handle:
```

| Access type | How the binary is obtained |
|---|---|
| `ociRegistry` | Image reference passed directly to scanner |
| `localBlob/v1` | Blob downloaded to a temporary file |
| `s3` | Tar archive downloaded and extracted |

### Dealing with Scanner Findings

Once a scanner produces findings, teams face two distinct problems:

* **True positives**: The package is genuinely vulnerable and must be addressed
* **False positives**: The scanner flags a package that is not actually exploitable or is already fixed

Both are common and both require different responses. The choice of scan input — binary vs. SBOM — affects which tools and workflows are available.

#### Patching true positives

> | Approach | How it works |

Reviewer: The number of columns is not consistent (ditto for the table below).

* **Reachability analysis** – some scanners can reduce false positives by checking whether a vulnerable code path is actually reachable in the scanned binary
    * To analyse reachability, the artefact must be present
    * Alternatively, VEX statements (e.g. `not_affected` with justification `vulnerable_code_not_in_execute_path`) can capture the same signal independently of scan input — a component author can ship VEX alongside the SBOM in the OCM component, and ODG can apply it during rescoring without re-scanning the binary

## Considered Options

### 1. Binary scanning only

The scanner extension fetches the OCM resource binary and passes it directly to the scanner. For `ociRegistry` resources the image reference is passed; for `localBlob/v1` and `s3` resources the binary is downloaded first.

If an SBOM is available for the resource, CVE findings from the binary scan could be
cross-checked against it: a finding for a package that does not appear in the SBOM may

> indicate a false positive (phantom detection) and can be flagged for review or suppressed (1)

Reviewer: Where does the `(1)` belong to?

| | Binary scanning only |
|---|---|

> | Upgrade dependency | ✓ detected |

Reviewer: The table options used in the `Considered Options` section do not fully match those defined in the `Context and Problem Statement` section (some are not covered).

| `localBlob/v1` | Blob downloaded to a temporary file |
| `s3` | Tar archive downloaded and extracted |

### Dealing with Scanner Findings

Once a scanner produces findings, teams face two distinct problems:

* **True positives**: The package is genuinely vulnerable and must be addressed
* **False positives**: The scanner flags a package that is not actually exploitable or is already fixed

Both are common and both require different responses. The choice of scan input — binary vs. SBOM — affects which tools and workflows are available.

#### Patching true positives

| Approach | How it works |
|---|---|
| Upgrade the dependency | Update version in source, rebuild, scanner sees new version |
| Backported fix with release suffix | Distro ships a patched build at the same upstream version but with an incremented release suffix (e.g. `curl-7.61.1-22.el8`) |
| In-place source patch | Plain-text diffs stored in a `patches/` folder in the repo (e.g. via [patch-package](https://github.com/ds300/patch-package)); applied during the build; upstream version number does not change; patches may be human-written, backported, or AI-assisted |

> | Direct image patching (e.g. [Copa](https://github.com/project-copacetic/copacetic), CNCF sandbox) | Reads a scan report, fetches updated packages from the distro repo via the package manager (`apt`, `pip`, `npm`...), and adds a patch layer to the existing image |

Reviewer 2: both patch options change the manifest digest and thus produce new artefacts.
how is this different from release a new version ocm artefact version from ODG perspective?

Author: Yes, if the dev team uses one of these approaches to fix the vulnerability findings in ODG, they produce a new artifact and digests change.

So if the CVE scanners still flag the CVE, the dev team must triage them in ODG manually, so additional work in ODG is necessary

Author: I also think having 5 would be great. However, there might also be scenarios where you don't want to do this, e.g. no access to container images and only want to scan SBOMs

Also reachability analysis is primarily a feature of commercial scanners

#### Suppressing false positives

| Approach | How it works |
|---|---|
| Fix the SBOM | Remove a phantom package or correct a misidentified PURL directly in the SBOM and ship the corrected SBOM in the OCM component |
| Reachability analysis | Statically analyse source or binary to determine whether the vulnerable function is actually reachable from the application's call graph |
| In-place patch not recognised | After a `patches/`-style fix the version is unchanged, so version-based scanners may still flag the CVE |
| VEX (e.g. [OpenVEX](https://github.com/openvex/spec) or CycloneDX VEX) | Structured attestation with status (`not_affected`, `fixed`, `under_investigation`) and a machine-readable justification (e.g. `vulnerable_code_not_in_execute_path`) |

> | Severity reranking | Override the scanner-assigned CVSS/severity with a context-specific score — e.g. downgrade a Critical to Low because the vulnerable service is internal-only, or because exploit complexity is high in the deployment context. Does not suppress the finding but reduces its priority in triage queues. |

Reviewer: > Does not suppress the finding but reduces its priority in triage queues.

ODG (and other tools) usually also feature to triage a finding fully.

Author: This is more about using CVSS to calculate a new numeric score by applying modifiers to the base score formula, e.g. https://www.first.org/cvss/calculator/3.1

This would happen before the ODG triage/rescoring, so findings would get a more accurate score before they land in triage/rescoring
